src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+    public NetworkElement updateNetworkElement(NetworkElement portGroup, List<NetworkElement> inspectionPorts)
             throws Exception {
-        // no-op
-        return null;
+
+        this.utils.throwExceptionIfNullElementAndId(portGroup, "Port Pair Group ServiceFunctionChain Id");
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "Port Pair Group update memeber list");
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(portGroup.getElementId());
+
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Update failed",
+                    portGroup.getElementId()));
+        }
+
+        sfc.getPortPairGroups().clear();
+        this.utils.addPortPairGroupToServiceFunction(inspectionPorts, sfc);
+
+        return this.txControl.required(() -> {
+            return this.em.merge(sfc);
+        });
     }
 
     @Override
-    public void deleteNetworkElement(NetworkElement portGroupId) throws Exception {
-        // no-op
+    public void deleteNetworkElement(NetworkElement serviceFunctionChainId) throws Exception {
+
+        this.utils.throwExceptionIfNullElementAndId(serviceFunctionChainId, "ServiceFunctionChain Id");
+
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(serviceFunctionChainId.getElementId());
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Delete failed",
+                    serviceFunctionChainId.getElementId()));
+        }
+
+        this.txControl.required(() -> {
+            //need to delete sfc if found in a transaction so fetch it again
+            ServiceFunctionChainEntity sfcToDel = this.em.find(ServiceFunctionChainEntity.class, serviceFunctionChainId.getElementId());
+            if (sfcToDel == null) {
+                throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Delete failed",
+                        serviceFunctionChainId.getElementId()));
+            }
+            for (PortPairGroupEntity ppg :sfcToDel.getPortPairGroups()) {
+                ppg.setServiceFunctionChain(null);
+                this.em.merge(ppg);
+            }
+            this.em.remove(sfcToDel);
+            return null;
+        });
     }
 
     @Override
     public List<NetworkElement> getNetworkElements(NetworkElement element) throws Exception {
-        return null;
+
+        this.utils.throwExceptionIfNullElementAndId(element, "ServiceFunctionChain Id");
+
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(element.getElementId());
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Get failed",
+                    element.getElementId()));
+        }
+        //do we need to have transaction for get?
+        return this.txControl.required(() -> {
```

Comment:
you can just return sfc.getPortPairGroups since ppg's are network elements

Reply:
Couldn't modify as it is giving a compilation error.. 
return sfc.getPortPairGroups() -- giving a compile error

Reply:
try new ArrayList<>(sfc.getPortPairGroups())

Reply:
Thanks that worked..fixed it